baremetal/metal3machine_manager.go

```diff
@@ -1338,15 +1335,58 @@ func (m *MachineManager) SetNodeProviderID(ctx context.Context, bmhID, providerI
 		m.Log.Info("Target node is not found, requeuing")
 		return &RequeueAfterError{RequeueAfter: requeueAfter}
 	}
+	namespace := m.Metal3Machine.GetNamespace()
+	m3mName := m.Metal3Machine.GetName()
+	bmhName, err := m.getBmhNameFromM3Machine()
+	if err != nil {
+		m.Log.Info("unable to retrieve BMH from Metal3Cluster")
+		return &RequeueAfterError{RequeueAfter: requeueAfter}
+	}
+	providerIDLegacy := *providerID
+	providerIDNew := fmt.Sprintf("metal3://%s/%s/%s", namespace, bmhName, m3mName)
+	if !m.Metal3Cluster.Spec.NoCloudProvider {
+		// Settig providerID on nodes takes time. Continuous monitoring is required.
+		nodesWithProviderIDset := 0
+		for _, node := range nodes.Items {
+			m.Log.Info("PP0: ", node.Spec.ProviderID)
+			if node.Spec.ProviderID == "" {
+				m.Log.Info("PP1: NoCloudProvider set to false. But, no providerID is set on the node either", node.GetName())
+				nodesWithProviderIDset++
+			} else if node.Spec.ProviderID == providerIDNew {
+				m.Log.Info("PP2: Found a node with matching node providerID in new format")
+				*providerID = providerIDNew
+				return nil
+			} else if node.Spec.ProviderID == providerIDLegacy {
+				m.Log.Info("PP3: Found a node with matching node providerID in legacy format")
+				*providerID = providerIDLegacy
+				return nil
+			}
+		}
+		// We have considered all nodes, and no matching node was found.
+		if nodesWithProviderIDset == nodesCount {
+			return errors.New("No matching node with the providerID found: ")
+		}
+		m.Log.Info("Waiting for more nodes to have their providerID property set before declaring failure")
+		return &RequeueAfterError{RequeueAfter: requeueAfter}
+	}
+	var nodeVar corev1.Node
 	for _, node := range nodes.Items {
```

Comment:
at this point you should check there is only one node (it should have used the label matching). If not, error out

Reply:
I think that is implicit in the logic, since it will return the moment it finds it.

Reply:
I have thought about it as well, but I was assuming the user is responsible for providing uniq nodes. But, if we are to verify as well, I should probably move the logic to a different function so that uniqueness is checked.

Reply:
There's no need to make it a whole thing, just a small length check is enough